contracts/src/Verifier.sol

```diff
@@ -73,101 +119,289 @@ contract Verifier is IVerifier, Ownable {
+        Hash _commitId = _generateCommitId(_modelContentId, _merkleRoot);
+        Hash _challenge = Hash.wrap(Challenge.generateChallenge(difficulty));
+        uint8 _difficulty = difficulty;
+
+        commits[_modelContentId].push(_commitId);
+        proverAddressToCommits[msg.sender].push(_commitId);
+        commitIdToCommit[_commitId] = Commit({
+            id: _commitId,
+            modelContentId: _modelContentId,
+            merkleRoot: _merkleRoot,
+            challenge: _challenge,
+            difficulty: _difficulty,
+            proverAddress: msg.sender,
+            isRevealed: false
+        });
+
+        emit CommitAdded(_commitId, msg.sender);
+
+        return (_commitId, _challenge, _difficulty);
+    }
+
+    function getCommit(
+        Hash _commitId
     )
         external
         view
-        validateLimitParameter(limit)
-        returns (ModelArrayElement[] memory paginatedModels)
+        checkIfCommitExists(_commitId)
+        onlyOwner
+        returns (Commit memory)
+    {
+        return commitIdToCommit[_commitId];
+    }
+
+    function getCommitsOfModel(
+        Hash _modelContentId,
+        uint32 _offset,
+        uint32 _limit
+    )
+        external
+        view
+        checkIfModelExists(_modelContentId)
+        validateOffsetParameter(_offset, commits[_modelContentId].length)
+        validateLimitParameter(_limit)
+        onlyOwner
+        returns (Hash[] memory)
+    {
+        return _paginateCommits(commits[_modelContentId], _offset, _limit);
+    }
+
+    function getCommitsOfProver(
+        address _proverAddress,
+        uint32 _offset,
+        uint32 _limit
+    )
+        external
+        view
+        validateOffsetParameter(
+            _offset,
+            proverAddressToCommits[_proverAddress].length
+        )
+        validateLimitParameter(_limit)
+        isValidProver(_proverAddress)
+        returns (Hash[] memory)
+    {
+        return
+            _paginateCommits(
+                proverAddressToCommits[_proverAddress],
+                _offset,
+                _limit
+            );
+    }
+
+    function updateChallenge(
+        Hash _commitId
+    )
+        external
+        checkIfCommitExists(_commitId)
+        isValidProver(commitIdToCommit[_commitId].proverAddress)
+        returns (Hash)
+    {
+        uint8 _difficulty = difficulty;
+        Hash _challenge = Hash.wrap(Challenge.generateChallenge(_difficulty));
+        commitIdToCommit[_commitId].challenge = _challenge;
+        commitIdToCommit[_commitId].difficulty = _difficulty;
+
+        emit ChallengeUpdated(_commitId, msg.sender);
+
+        return _challenge;
+    }
+
+    function getDifficulty() external view onlyOwner returns (uint8) {
+        return difficulty;
+    }
+
+    function updateDifficulty(
+        uint8 _difficulty
+    ) external isValidDifficulty(_difficulty) onlyOwner {
+        difficulty = _difficulty;
+    }
+
+    function reveal(
```

Comment:
This function does the following verifications in this order.
1. Verify that all of the given leaves contain the generated challenge.
2. Verify the given Merkle proofs (i.e. `proofs`, `proofFlags`, and `leaves`).